File: userService/src/main/java/dev/whitechoke/userService/domain/ProfileService.java
package dev.whitechoke.userService.domain;

import dev.whitechoke.commonLibs.http.FormResponseDto;
import dev.whitechoke.commonLibs.http.UserPreferencesResponseDto;
import dev.whitechoke.commonLibs.kafka.ProfileCreatedEvent;
import dev.whitechoke.commonLibs.kafka.ProfileDeactivateEvent;
import dev.whitechoke.userService.api.dto.ProfileCreateRequestDto;
import dev.whitechoke.commonLibs.http.ProfileGetByFilterRequestDto;
import dev.whitechoke.commonLibs.http.ProfileResponseDto;
import dev.whitechoke.userService.domain.db.ProfileRepository;
import dev.whitechoke.userService.utils.ProfileMapper;
import dev.whitechoke.userService.utils.ProfileValidator;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class ProfileService {

    private final ProfileRepository profileRepository;
    private final ProfileValidator validator;
    private final ProfileMapper mapper;
    private final ApplicationEventPublisher eventPublisher;

    @Transactional
    @CachePut(value = "profile", key = "#result.id")
    public ProfileResponseDto createProfile(
            ProfileCreateRequestDto request
    ) {
       validator.validateCreateRequest(request);
       var entity = mapper.toProfileEntity(request);

       entity.setRegisteredAt(Instant.now());
       var saved = profileRepository.save(entity);

       var event = ProfileCreatedEvent.builder()
               .telegramId(saved.getTelegramId())
               .latitude(saved.getCoordinates().getY())
               .longitude(saved.getCoordinates().getX())
               .maxAge(saved.getMaxAge())
               .minAge(saved.getMinAge())
               .radius(saved.getSearchRadius())
               .gender(saved.getSearchGender())
               .build();

        eventPublisher.publishEvent(event);

       log.info("Created profile with id={}", saved.getId());
       return mapper.toResponseDto(saved);
    }

    @Cacheable(value = "profile", key = "#id")
    public ProfileResponseDto getProfileById(Long id) {
        var entity = profileRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Not found entity with id=" + id));

        return mapper.toResponseDto(entity);
    }

    public List<Long> getProfilesByFilter(
            ProfileGetByFilterRequestDto request
    ) {
        return  profileRepository.findProfilesByFilter(
                request.latitude(),
                request.longitude(),
                request.searchRadius(),
                request.maxAge(),
                request.minAge(),
                request.gender()
        );
    }

    @Transactional
    @CacheEvict(value = "profile", key = "#result")
    public Long deactivateProfile(Long telegramId) {
        var entity = profileRepository.findByTelegramId(telegramId)
                .orElseThrow(() -> new EntityNotFoundException("Not found entity with telegram id=" + telegramId));

        profileRepository.delete(entity);
        
        var event = ProfileDeactivateEvent.builder()
                .id(entity.getId())
                .telegramId(entity.getTelegramId())
                .build();
        
        eventPublisher.publishEvent(event);

        log.info("Deactivated profile with telegram id={}", telegramId);

        return entity.getId();
    }

    public UserPreferencesResponseDto getUserPreferences(Long telegramId) {
        var profile = profileRepository.findByTelegramId(telegramId)
                .orElseThrow(() -> new EntityNotFoundException("Not found entity with id=" + telegramId));

        return UserPreferencesResponseDto.builder()
                .latitude(profile.getCoordinates().getY())
                .longitude(profile.getCoordinates().getX())
                .maxAge(profile.getMaxAge())
                .minAge(profile.getMinAge())
                .searchGender(profile.getSearchGender())
                .searchRadius(profile.getSearchRadius())
                .build();
    }

    public FormResponseDto getForm(Long id) {
        var profile = profileRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Not found entity with id=" + id));

        return FormResponseDto.builder()
                .id(profile.getId())
                .telegramId(profile.getTelegramId())
                .name(profile.getName())
                .birthday(profile.getBirthday())
                .gender(profile.getGender())
                .bio(profile.getBio())
                .build();
    }
}

File: bot/app/api/profile_service.py
import aiohttp

genders = {
    "Мужчина": "MALE",
    "Женщина": "FEMALE"
}

service_url = "http://localhost:8080/api/v1/profiles"

async def create_profile(tg_id: int, request: dict) -> None:
    request_body = {
        "telegramId": tg_id,
        "name": request["name"],
        "birthday": request["birthday"],
        "maxAge": request["maxAge"],
        "minAge": request["minAge"],
        "gender": genders[request["gender"]],
        "searchGender": genders[request["searchGender"]],
        "latitude": request["coordinates"]["latitude"],
        "longitude": request["coordinates"]["longitude"],
        "bio": request["bio"],
        "searchRadius": request["searchRadius"],
        "photoLink": request["photoLink"],
    }
    
    async with aiohttp.ClientSession() as session:
        await session.post(url=service_url, json=request_body)

async def get_profile(tg_id: int) -> aiohttp.ClientResponse:
    async with aiohttp.ClientSession() as session:
        response = await session.get(f"{service_url}/tg/{tg_id}")
        await response.json()
        
        return response
    
async def activate_profile(tg_id: int) -> int:
    async with aiohttp.ClientSession() as session:
        response = await session.patch(f"{service_url}/activate/{tg_id}")
        
        return response.status
    
async def deactivate_profile(tg_id: int) -> int:
    async with aiohttp.ClientSession() as session:
        response = await session.delete(f"{service_url}/{tg_id}")
        
        return response.status
    

